### PR TITLE
rafthttp: expose struct to set configuration

### LINKED ISSUE #3568: test etcd snapshot send in limited network bandwidth
If the network bandwidth between etcd members is limited, and cannot send message snapshot before reading header timeout(5s), it is possible that the follower cannot receive any valid snapshot forever in this scenario.

After testing, if this is the case, it should be fixed.

### PR DESCRIPTION
**Pull request by yichengq**
transport takes too many arguments and the new function is unable to
read. Change the way to set fields in transport struct directly.

### CODE REVIEW

**Comment by xiang90** [comment]
Let us do one thing at a time. The first commit looks good to me.

**Comment by xiang90** [comment]
We should open another pr for the 2nd and 3rd commits.

**Comment by yichengq** [comment]
Addressed the comments. Ready for the next round.

> Let us do one thing at a time.

ok.

File: etcdserver/server_test.go [hunk 1]
```diff
@@ -1468,6 +1468,7 @@ func (n *readyNode) Ready() <-chan raft.Ready { return n.readyc }
 
 type nopTransporter struct{}
 
+func (s *nopTransporter) Start()                              { return }
```
Comment (xiang90) [inline etcdserver/server_test.go]: remove return. just {}

Comment (yichengq) [inline etcdserver/server_test.go]: done.

File: rafthttp/transport.go [hunk 2]
```diff
@@ -78,13 +81,24 @@ type Transporter interface {
...
 }
 
-type transport struct {
-	roundTripper http.RoundTripper
-	id           types.ID
-	clusterID    types.ID
-	raft         Raft
-	serverStats  *stats.ServerStats
-	leaderStats  *stats.LeaderStats
+// Transport is the implementation of Transporter. It provides the functionality
```
Comment (xiang90) [inline rafthttp/transport.go]: Transport implements Transporter interface.

File: rafthttp/transport.go [hunk 3]
```diff
@@ -78,13 +81,24 @@ type Transporter interface {
...
+// to send raft messages to peers, and receive raft messages from peers.
+// The Handler in Transport needs to serve requests sent to peer URLs of the
```
Comment (xiang90) [inline rafthttp/transport.go]: I cannot really understand line 86 and line 87. I am not sure which Handler(There is not handler field in Transport struct) are you referring to and I am not sure what does "work well" really mean. Should I provide a Handler to Transport? Or is it an implementation detail that I should not care about?

File: rafthttp/transport.go [hunk 4]
```diff
@@ -78,13 +81,24 @@ type Transporter interface {
...
+// local member for remote Transport to work well.
+// Transport needs to call Start before calling other functions, and call
```
Comment (xiang90) [inline rafthttp/transport.go]: Transport needs to call Start? Or the user of transport needs to call Start? From the code, Start method is called by the user of Transport and never be called by Transport itself.

Comment (yichengq) [inline rafthttp/transport.go]: user needs to call Start. will fix.

File: rafthttp/transport.go [hunk 5]
```diff
@@ -78,13 +81,24 @@ type Transporter interface {
...
+// Stop when it is no longer used.
+type Transport struct {
+	RoundTripper http.RoundTripper  // roundTripper to send requests
+	ID           types.ID           // local member ID
+	ClusterID    types.ID           // raft cluster ID for request validation
+	Raft         Raft               // raft state machine, to which the Transport processes received messages and reports status
```
Comment (xiang90) [inline rafthttp/transport.go]: Transport processes received message to raft state machine? The grammar seems to be wrong.

File: rafthttp/transport.go [hunk 6]
```diff
@@ -78,13 +81,24 @@ type Transporter interface {
...
+	ServerStats  *stats.ServerStats // used to record general communication statistics
```
Comment (xiang90) [inline rafthttp/transport.go]: communication -> transportation. We do not want to introduce new term when we already have one.

File: rafthttp/transport.go (hunk 3 above)
Comment (yichengq) [inline rafthttp/transport.go]: The comment should clarify the handler entity better. Changed it to:

```
// User should let the handler returned by Handler method in Transport
// serve requests sent to peer URLs of the local member, which helps remote peers
// to connect with local member.
```

File: rafthttp/transport.go (hunk 5 above)
Comment (yichengq) [inline rafthttp/transport.go]: raft should be the entity to process the message. change the verb to `forward` to make it readable.

File: rafthttp/transport.go (hunk 6 above)
Comment (yichengq) [inline rafthttp/transport.go]: ok

File: rafthttp/transport.go [hunk 7]
```diff
@@ -78,13 +81,25 @@ type Transporter interface {
...
-type transport struct {
-	roundTripper http.RoundTripper
-	id           types.ID
-	clusterID    types.ID
-	raft         Raft
-	serverStats  *stats.ServerStats
-	leaderStats  *stats.LeaderStats
+// Transport implements Transporter interface. It provides the functionality
+// to send raft messages to peers, and receive raft messages from peers.
+// User should let the handler returned by Handler method in Transport
```
Comment (xiang90) [inline rafthttp/transport.go]: User should call Handler method to get a handler to serve requests received from peerURLs. (remove the unnecessary details.)

File: rafthttp/transport.go [hunk 8]
```diff
@@ -78,13 +81,25 @@ type Transporter interface {
...
+// remote peers to connect with local member.
+// User needs to call Start before calling other functions, and call
+// Stop when the Transport is no longer used.
+type Transport struct {
+	RoundTripper http.RoundTripper  // roundTripper to send requests
+	ID           types.ID           // local member ID
+	ClusterID    types.ID           // raft cluster ID for request validation
+	Raft         Raft               // raft state machine, to which the Transport forwards received messages and reports status
+	ServerStats  *stats.ServerStats // used to record general transportation statistics
+	LeaderStats  *stats.LeaderStats // used to record transportation statistics with followers when performing as leader in raft protocol
```
Comment (xiang90) [inline rafthttp/transport.go]: The comment is too long. I would suggest to move the comment on top of the field instead of in one line.

**Comment by xiang90** [comment]
LGTM

File: rafthttp/transport.go (hunk 7 above)
Comment (yichengq) [inline rafthttp/transport.go]: ok. i will try to remove unnecessary details in the future.

**Pull request merged**